File: 2216-delete-the-middle-node-of-a-linked-list/delete-the-middle-node-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteMiddle(ListNode head) {
        if(head==null || head.next==null) return null;
        ListNode prev=null, slow = head, fast = head;

        while(fast!=null && fast.next!=null){
            prev=slow;
            slow=slow.next;
            fast=fast.next.next;
        }
        //slow will point to the node to be deleted and prev is the node before it
        if(slow!=head) prev.next = slow.next;
        else head = null;

        slow.next=null;
        // delete slow;
        return head;
    }
}

File: 2216-delete-the-middle-node-of-a-linked-list/delete-the-middle-node-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        // Method 1 [BruteForce] O(2N), O(1)
        if(head==NULL || head->next==NULL) return NULL;
        int cnt = 0;
        ListNode* temp = head;
        while(temp!=NULL){
            cnt++;
            temp = temp->next;
        }

        int res = cnt/2;
        temp = head;
        while(temp!=NULL){
            res--;
            if(res==0) {
                ListNode* middle = temp->next;
                temp->next = temp->next->next;
                middle->next = NULL;
                delete middle;
                break;
            }
            temp = temp->next;
        }
        return head;

        // Method 2 [Optimised] O(N), O(1)
        // if(head==NULL || head->next==NULL) return NULL;
        // ListNode *prev=NULL, *slow = head, *fast = head;

        // while(fast!=NULL && fast->next!=NULL){
        //     prev=slow;
        //     slow=slow->next;
        //     fast=fast->next->next;
        // }
        // //slow will point to the node to be deleted and prev is the node before it
        // if(slow!=head) prev->next = slow->next;
        // else head = nullptr;

        // slow->next=NULL;
        // delete slow;
        // return head;
    }
};